pkg/controller/direct/certificatemanager/dnsauthorization_controller.go

```diff
@@ -0,0 +1,334 @@
+
+	// Get location
+	// TODO: Add location when field is added
+	// location := obj.Spec.Location
+	location := "global"
+
+	var id *CertificateManagerDNSAuthorizationIdentity
+	// TODO: Add ExternalRef when field is added
+	// externalRef := direct.ValueOf(obj.Status.ExternalRef)
+	externalRef := ""
+	if externalRef == "" {
+		id = BuildID(projectID, location, resourceID)
+	} else {
+		id, err = asID(externalRef)
+		if err != nil {
+			return nil, err
+		}
+
+		if id.Parent.Project != projectID {
+			return nil, fmt.Errorf("CertificateManagerDNSAuthorization %s/%s has spec.projectRef changed, expect %s, got %s",
+				u.GetNamespace(), u.GetName(), id.Parent.Project, projectID)
+		}
+		if id.Parent.Location != location {
+			return nil, fmt.Errorf("CertificateManagerDNSAuthorization %s/%s has spec.location changed, expect %s, got %s",
+				u.GetNamespace(), u.GetName(), id.Parent.Location, location)
+		}
+		if id.DnsAuthorization != resourceID {
+			return nil, fmt.Errorf("CertificateManagerDNSAuthorization  %s/%s has metadata.name or spec.resourceID changed, expect %s, got %s",
+				u.GetNamespace(), u.GetName(), id.DnsAuthorization, resourceID)
+		}
+	}
+
+	// Get certificatemanager GCP client
+	gcpClient, err := m.client(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &Adapter{
+		id:        id,
+		gcpClient: gcpClient,
+		desired:   obj,
+	}, nil
+}
+
+func (m *model) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type Adapter struct {
+	id        *CertificateManagerDNSAuthorizationIdentity
+	gcpClient *gcp.Client
+	desired   *krm.CertificateManagerDNSAuthorization
+	actual    *certificatemanagerpb.DnsAuthorization
+}
+
+var _ directbase.Adapter = &Adapter{}
+
+func (a *Adapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("getting CertificateManagerDNSAuthorization", "name", a.id.FullyQualifiedName())
+
+	req := &certificatemanagerpb.GetDnsAuthorizationRequest{Name: a.id.FullyQualifiedName()}
+	dnsauthorizationpb, err := a.gcpClient.GetDnsAuthorization(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting CertificateManagerDNSAuthorization %q: %w", a.id.FullyQualifiedName(), err)
+	}
+
+	a.actual = dnsauthorizationpb
+	return true, nil
+}
+
+func (a *Adapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	u := createOp.GetUnstructured()
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("creating DnsAuthorization", "name", a.id.FullyQualifiedName())
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	resource := CertificateManagerDNSAuthorizationSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	resource.Labels = make(map[string]string)
+	for k, v := range a.desired.GetObjectMeta().GetLabels() {
+		resource.Labels[k] = v
```

Comment:
I think you may need to add back the `managed-by` label since this resource is already in Beta.